// projects/file.js
let project = require("projects");
const FileTree = require("filetree");

class ProjectFile{
    constructor(name){
        this.setup();
        const that = this;

        // load the file
        project.load(name).then(function(data){
            that.setFile(data);

        }).catch(function(){
            // then this is a new file
            console.log("New File");
            that.file.name = name;
            that.save();
        });
        console.log(this);
    }

    onLoad(){
        let projectEl = document.getElementById("project-list-current");

        projectEl.innerText = this.file.name;
    }

    setFile(file){
        this.file = file;

        this.fileTrees = {};

        let element = document.getElementById("files-list-menu");
        element.innerHTML = "";

        for(let d in this.file.data){
            let options = {
                name: d
            };

            this.fileTrees[d] = new FileTree(
                element, 
                this.file.data[d].files, 
                options
            );

        }

        this.onLoad();
    }

    setup(){
        this.file = {
            name: "",
            namespace: "",
            data: {
                advancements:{},
                functions:{},
                loot_tables:{},
                structures:{},
                recipes:{},
                tags:{}
            }
        };
    }

    setData(data){
        for(let d in data){
            if(d === "data"){
                continue;
            }
            this.file[d] = data[d];
        }

        return this;
    }

    /**
     * save the data
     */
    save(){
        console.log("Save File", this.file);
        project.save(this.file);
    }

    get data(){
        return this.file;
    }

    /**
     * 
     */
    updateDOM(){

    }
}

module.exports = ProjectFile;
