// projects/settings.js
const jsEvents = require("jsevents");

const Modal = require("modal");


class ProjectFileSettings {

    constructor(that){
        this.events = new jsEvents(this);
        this.parent = that;

        this.open();
    }

    /**
     * open a settings dialog
     */
    open(){
        const that = this;
        let content = document.createChild("p", {
            text: "Set some settings"
        });

        content.appendChild(this.deleteButton);

        let modal = new Modal(this.parent.name + " settings", content, "Save");
        console.log(modal);
        modal.on("cancel", function(){
            this.destroy();
            that.modal = null;
        });

        modal.on("success", function(){
            console.log("Submitted");
            this.destroy();
            that.modal = null;
        });

        this.modal = modal;
    }

    close(){
        if(this.modal){
            this.modal.destroy();
        }
    }

    get deleteButton(){
        const that = this;
        let button = document.createChild("button", {
            classList: ["btn", "btn-danger", "float-right"],
            type: "button",
            text: "Delete",
            on :{
                click: function(){
                    that.events.triggerCallback("delete", that, []);
                }
            }
        });

        button.insertBefore(document.createChild("i", {
            classList: ["material-icons", "align-middle"],
            text: "delete"
        }), button.firstChild);

        return button;
    }
}

module.exports = ProjectFileSettings;
