// blockly/blockly.js
// load all of the blocks
console.log("Load Blockly");
require("./generators/functions.js");

// require("./blocks/string.js");
// require("./blocks/say.js");
require("./loadAllBlocks.js");

const FunctionsBlockly = require("./functions");

const blocklyDiv = document.getElementById('content');
module.exports.blocklyDiv = blocklyDiv;

/**
 * get the blockly div
 */
module.exports.getDiv = function(){
    return document.getElementById("content");
};

/**
 * the main workspace varaible
 */
let workspace = null;
module.exports.workspace = workspace;
window.workspace = workspace;

/**
 * 
 */
module.exports.start = function () {

    workspace = Blockly.inject(blocklyDiv,
        { toolbox: document.getElementById('toolbox') });

    function myUpdateFunction(event) {
        var code = Blockly.Functions.workspaceToCode(workspace);
        console.log("%c---------------------", "color: grey;");
        console.log(code);
        console.log("%c---------------------", "color: grey;");

    }
    workspace.addChangeListener(myUpdateFunction);
};

/**
 * Open a file in blockly
 * 
 * @param {String} name 
 * @param {Object} file 
 */
module.exports.open = function (name, file) {
    if(workspace){
        workspace.close();
    }
    console.log("%cOpen blockly", "color: green;", name);
    switch (name) {
        case "functions":
            workspace = new FunctionsBlockly(file);
            break;
        default:
    }
    console.log("Workspace", workspace);
    return workspace;
};
